docs/internal/GeoffsRecoilAssessment.md

```diff
@@ -0,0 +1,176 @@
+
+Finally, I rewrote the application to use Recoil. I flattened the application state into atoms. Because relationships between the data create a logical hierarchy, I wrote utility functions to make cross referencing between data islands easy. I then wrote selectors to shape the data for my React components. The container components leveraged useRecoilValue and useRecoilState to respond to state changes. Because the same application state change be updated from different React components, I implemented a dispatcher pattern based on useRecoilCallback.  I had several missteps as I learned how Recoil works, and discovered how my application was affected by flattening the state.
+
+# Recoil API: Pros, Cons, Unknowns, and Hazards
+
+## Overall
+- PRO: Built for hooks from the start and follows React concepts
+- PRO: Atom and selectors allow for composing applications from small pieces without introducing unnecessary dependencies.
+- PRO: Significantly reduces the complexity and code volume in components.
+- PRO: Looking toward the future of asynchronous, multi-threaded Javascript.
+- PRO: Recoil API surface is tiny
+- PRO: Lightning fast (at least for my application size)
+- CON: Still experimental, few releases, limited contributes - not yet mature or fully tested.
+
+## Atoms
+- PRO: Like their name, atoms let you model your data as small and independent plain types.
+- PRO: Serializing atoms to storage is straightforward compared to trying to rehydrate an observed state tree.
+- PRO: The default state of an atom with the reset() hook is convenient for clearing application state.
+- CON: Atoms are often arrays of items which means needing to look up items. This requires more undefined checks for cases where items are not found.
+- CON: There is additional "param baggage" to pass multiple atoms to different functions vs. passing the root of the application state or a child.
+- HAZARD: Care has to be taken to avoid updating atom state by mutating the current state.  Only a set will trigger change detection.
+
+## Selectors
+- PRO: Simple to shape data for React components. 
+- PRO: Like atoms they are small and can be built as an library of independent data shapes.
+- PRO: The dependency chain from one selector to other selectors or atoms is easy to follow.
+- PRO: Less ceremony compared to Redux's mapStateToProps.
+- UNKNOWN: Unsure if a selector would re-compute its value if no UI is subscribed to it currently but it had been subscribed before.
+- UNKNOWN: The writable selector is not well documented and unclear when to use.
```

Comment:
A selector is a derivative obtained from atoms. This should never be writable right?

Reply:
Let's call this out as an anti-pattern for now.